### ISSUE TITLE
attention issue

### ISSUE THREAD

**Issue by nanshen308** [DL4mHealth/Medformer#7]
Hello author, I have a question to consult:
In the paper, the shape of attention z within the granularity is different from that of x, while x_in is used in the code,

![1](https://github.com/user-attachments/assets/e3741f61-1cd1-4f2c-9be6-22cdc00a4cc3)

 def forward(self, x, attn_mask=None, tau=None, delta=None):
        attn_mask = attn_mask or ([None] * len(x))
        # Intra attention
        x_intra = []
        attn_out = []
        for x_in, layer, mask in zip(x, self.intra_attentions, attn_mask):
            _x_out, _attn = layer(x_in, x_in, x_in, attn_mask=mask, tau=tau, delta=delta)
            x_intra.append(_x_out)  # (B, Li, D)
            attn_out.append(_attn)
        if self.inter_attention is not None:
            # Inter attention
            routers = torch.cat([x[:, -1:] for x in x_intra], dim=1)  # (B, N, D)
            x_inter, attn_inter = self.inter_attention(
                routers, routers, routers, attn_mask=None, tau=tau, delta=delta
            )
            x_out = [
                torch.cat([x[:, :-1], x_inter[:, i : i + 1]], dim=1)  # (B, Li, D)
                for i, x in enumerate(x_intra)
            ]
            attn_out += [attn_inter]
        else:
            x_out = x_intra
        return x_out, attn_out

Looking forward to your reply

**Comment by YiheWang** [comment]
For simplicity, we use the last column of x as the router, which differs from the paper's description of creating a separate router and concatenating it. While this approach results in a single-column difference from the paper, it significantly simplifies the implementation.

**Comment by YiheWang** [comment]
One thing I forgot to mention, since we do padding to add empty time points with the length of the patch to avoid patches not being divided, the last patch is basically the same as the router initialization.

line 253 in Embed.py:
`self.paddings = [nn.ReplicationPad1d((0, stride)) for stride in stride_list]`

**Comment by nanshen308** [comment]
Hello author, in your code, for the intra-granular attention, you pay attention to xi directly. Then, in the inter-granular attention, you perform self-attention on ui, update it, and then replace the last column of xi with ui. Therefore, in your code, the final xi after the attention module contains the router. In the paper, your final output is also xi, but this xi does not contain the router, only zi contains the router. The paper classifies xi, not zi.

**Comment by nanshen308** [comment]
Hello author,
(1) In your code, for the intra-granular attention, you pay attention to xi directly. Then, in the inter-granular attention, you perform self-attention on ui, update it, and then replace the last column of xi with ui. Therefore, in your code, the final xi after the attention module contains the router. In the paper, your final output is also xi, but this xi does not contain the router, only zi contains the router. The paper classifies xi, not zi.
(2) In the intra-granularity, the paper uses two attentions to update xi and ui. This is actually self-attention on zi. Why not just write zi<--atten(zi,zi,zi)

**Comment by YiheWang** [comment]
Hi, thanks for the suggestion. Yes, it might be a better way to write zi<--atten(zi,zi,zi). As we said in the paper, the updating for xi and ui is done concurrently. The reason for the two lines of attention formulas is to demonstrate better that both xi and routers are updated.

**Comment by nanshen308** [comment]
Hello author,
In the paper, the input to the classification head is xi. However, it seems more reasonable to input the classification head to zi.
Thank you very much for the author's careful guidance

**Comment by YiheWang** [comment]
Yes, actually, we did some post-experiments after the paper submission, and we found the performance can be improved in many cases by only input ui, which is routers (the last column of xi in code implementation). But we don't want to run all the experiments again and update the tables. So we keep it and may update it in the future works.

**Comment by nanshen308** [comment]
I understand.Thank you very much.
